app/javascript/Domain/UseCases/Nations/Bonds/AffordableBonds.js

```diff
@@ -0,0 +1,37 @@
+export default class AffordableBonds {
+  static* bondsPurchasableInFull(nation, investor) {
+    for (const bond of nation.unsoldBondsByInterestValue.values()) {
+      if (investor.cash >= bond.cost) {
+        yield bond;
+      }
+    }
+  }
+  static* allAffordableBonds(nation, investor) {
+    // TODO: refactor player entity to allow quick filtering of bonds by nation
+    const sameNationBonds = [];
+    for (const bond of investor.bonds) {
+      if (bond.nation === nation) {
+        sameNationBonds.push(bond);
+      }
+    }
+
+    for (const bond of nation.unsoldBondsByInterestValue.values()) {
+      if (investor.cash >= bond.cost) {
+        yield bond;
```

Comment:
Mostly to avoid having to overthink the logic on how to make sure we're not yielding duplicate bonds

In other to avoid duplicates, even if we yield those in `bondsPurchasableInFull` directly, we'd still have the wrap the logic below (the upgrade logic) in an `if (cannot purchase in full)`, which is about the same

```
for (const bond of bondsPurchasableInFull(nation, investor)) {
  yield bond;
}

for (const bond of nation.unsoldBondsByInterestValue.values()) {
  if (investor.cash < bond.cost) {
    // yield upgradable bonds logic
  }
}
```

vs
 
```
for (const bond of nation.unsoldBondsByInterestValue.values()) {
  if (investor.cash < bond.cost) {
    // yield upgradable bonds logic
  } else {
    // yield bond directly
  }
}
```

up to you idrc either way